File: 第四章_队列和栈/队列和栈(1)/Solution.java
import java.util.Stack;

public class Solution {

	Stack<Integer> stack_data = new Stack<>();// 存放数据的栈
	Stack<Integer> stack_min = new Stack<>();// 存放最小值的栈

	public void push(int node) {
		stack_data.push(node);
		if (stack_min.isEmpty() || node < stack_min.peek()) {// 当要插入的元素小于最小值栈顶元素时才执行插入或者栈为空
			stack_min.push(node);
		}
	}

	public void pop() {
		if (!stack_data.empty()) {// 当栈不为空时
			int node = stack_data.peek();
			if (node <= stack_min.peek()) {// 若弹出的元素小于等于最小值栈顶元素时，弹出最小值栈顶元素。
				stack_min.pop();
			}
			stack_data.pop();
		}
	}

	public int top() {
		return stack_data.peek();
	}

	public int min() {
		return stack_min.peek();
	}
}

File: 第四章_队列和栈/队列和栈(1)/Solution.cpp
class Solution {//这里采用的和JAVA的实现方法有一点不一样，最小值栈中会插入和数据栈中一样多的元素
public:
    stack<int>stack_data;//数据栈
    stack<int>stack_min;//最小值栈
          
    void push(int value) {
        stack_data.push(value);
        if(stack_min.empty()||value<stack_min.top())//当前元素小于最小值栈顶元素时插入当前元素，否则插入最小值栈顶元素(这里有不同)
            {
            stack_min.push(value);
        }
        else stack_min.push(stack_min.top());
    }
    void pop() {
        if(!stack_data.empty()){//弹出元素，不同判断直接一起弹出(这里有不同)
            stack_data.pop();
            stack_min.pop();
        }
    }
    int top() {
        return stack_data.top();
    }
    int min() {
    return stack_min.top();
    }
};